src/python-common/ceph/deployment/service_spec.py

```diff
@@ -108,27 +108,32 @@ class PlacementSpec(object):
     For APIs that need to specify a host subset
     """
 
-    def __init__(self, label=None, hosts=None, count=None, all_hosts=False, host_pattern=None):
-        # type: (Optional[str], Optional[List], Optional[int], bool, Optional[str]) -> None
-        if all_hosts and (count or hosts or label):
-            raise ServiceSpecValidationError('cannot combine all:true and count|hosts|label')
+    def __init__(self,
+                 label=None,  # type: Optional[str]
+                 hosts=None,  # type: Union[List[str],List[HostPlacementSpec]]
+                 count=None,  # type: Optional[int]
+                 host_pattern=None  # type: Optional[str]
```

Comment:
should this been ``Optional[List[str]]`` so you can do something like ``hostfoo*,hostbar*``?

Reply:
Though about that, too. But this is problematic in combination with `HostPlacementSpec`:

`hostfoo*,hostbar=a` would mean that we have to combine the pattern with the `HostPlacementSpec`s.

Reply:
host list and host_pattern list can be mutually exclusive?

Reply:
right now, they are mutually exclusive.